commit 1e63ff0c37e40816639806088f961208a4fed823
Author: Eric Anholt <eric@anholt.net>
Date:   Thu Nov 3 18:53:10 2016 -0700

    drm/vc4: Add support for rendering with ETC1 textures.
    
    The validation for it ends up being quite simple, but I hadn't got
    around to it before merging the driver.  For backwards compatibility,
    we also need to add a flag so that the userspace GL driver can easily
    tell if the kernel will allow ETC1 textures (on an old kernel, it will
    continue to convert to RGBA8)
    
    Signed-off-by: Eric Anholt <eric@anholt.net>
    (cherry picked from commit 7154d76fedf549607afbc0d13db9aaf02da5cebf)

Index: linux-4.9.47/drivers/gpu/drm/vc4/vc4_drv.c
===================================================================
--- linux-4.9.47.orig/drivers/gpu/drm/vc4/vc4_drv.c
+++ linux-4.9.47/drivers/gpu/drm/vc4/vc4_drv.c
@@ -81,6 +81,7 @@ static int vc4_get_param_ioctl(struct dr
 		pm_runtime_put_autosuspend(&vc4->v3d->pdev->dev);
 		break;
 	case DRM_VC4_PARAM_SUPPORTS_BRANCHES:
+	case DRM_VC4_PARAM_SUPPORTS_ETC1:
 		args->value = true;
 		break;
 	default:
Index: linux-4.9.47/drivers/gpu/drm/vc4/vc4_validate.c
===================================================================
--- linux-4.9.47.orig/drivers/gpu/drm/vc4/vc4_validate.c
+++ linux-4.9.47/drivers/gpu/drm/vc4/vc4_validate.c
@@ -644,6 +644,13 @@ reloc_tex(struct vc4_exec_info *exec,
 		cpp = 1;
 		break;
 	case VC4_TEXTURE_TYPE_ETC1:
+		/* ETC1 is arranged as 64-bit blocks, where each block is 4x4
+		 * pixels.
+		 */
+		cpp = 8;
+		width = (width + 3) >> 2;
+		height = (height + 3) >> 2;
+		break;
 	case VC4_TEXTURE_TYPE_BW1:
 	case VC4_TEXTURE_TYPE_A4:
 	case VC4_TEXTURE_TYPE_A1:
Index: linux-4.9.47/include/uapi/drm/vc4_drm.h
===================================================================
--- linux-4.9.47.orig/include/uapi/drm/vc4_drm.h
+++ linux-4.9.47/include/uapi/drm/vc4_drm.h
@@ -286,6 +286,7 @@ struct drm_vc4_get_hang_state {
 #define DRM_VC4_PARAM_V3D_IDENT1		1
 #define DRM_VC4_PARAM_V3D_IDENT2		2
 #define DRM_VC4_PARAM_SUPPORTS_BRANCHES		3
+#define DRM_VC4_PARAM_SUPPORTS_ETC1		4
 
 struct drm_vc4_get_param {
 	__u32 param;
Index: linux-4.9.47/dummy/rpi_1162_1e63ff0c37e40816639806088f961208a4fed823.txt
===================================================================
--- /dev/null
+++ linux-4.9.47/dummy/rpi_1162_1e63ff0c37e40816639806088f961208a4fed823.txt
@@ -0,0 +1 @@
+dummy file to ensure patch has content.
